Restructure project to become a full application

dfialho opened the pull request:
The project was re-structured to follow the python standards and **pip** can now be used to install the application.

Commit:
Move all code inside the 'mutual_debt' package

Commit:
Move main to its own file and add CLI

The data file used in the computations is now an input argument od the
application.

This commit also includes a new dependency 'docopt' to implement the
command line interface.

Commit:
Delete files not necessary for the application

This file is not needed for the application itself. It is just an
example file. The same is true for all media files inside the 'media'
directory.

Commit:
Add a setup.py file

Commit:
Include data file example

Commit:
Revert "Delete files not necessary for the application"

This reverts commit 9f40161

Commit:
Add images missing from README

Commit:
Fix import errors realted to missing 'mutual_debt' module

Commit:
Rename package to 'simplify-debts'

Commit:
Remove docopt as a dependency

Since the CLI is some simple, it can be implemented manually.

davidfialho14 commented:
The version number included in the setup.py file needs to be changed to the version number of the current or next release.

ric2b commented on mutual_debt/main.py:
```diff
@@ -0,0 +1,51 @@
+        print(__doc__)
+        return
+
+    data_file = sys.argv[1]
```
I think a default value would be nice so that you don't have to keep typing the file name if you use the default filename.

dfialho replied:
Ok, then what should be the default name?
What do you think about *debts.json*?

ric2b commented on mutual_debt/main.py:
```diff
@@ -0,0 +1,51 @@
+def main():
+
+    if len(sys.argv) != 2:
+        print(__doc__)
```
Nice.

dfialho replied:
I definitely agree with this one :smile:

ric2b commented on mutual_debt/main.py:
```diff
@@ -0,0 +1,51 @@
+    draw_graph(simplified_debt_graph, 'Simplified_Mutual_Debt')
+
+
+if __name__ == '__main__':
```
There's no reason to import this file, so why define main() instead of just running the code?

dfialho replied:
I did not understand what you meant "import this file". Are you talking about main.py? If so, then where is this file imported?

There are three reasons for using the main() function:
  1. it is standard way to do it in python;
  2. it avoids problems with shadowing names;
  3. and most important, to define a console script you have to point to a function. If you look in the setup.py file, you'll see that the console script 'simplify-debts' takes a path to a function, the main() function. As far as I know this is the only way to do it.

ric2b replied:
I meant that this isn't imported, so there was no need for the `if __name__ ==  '__main__'` technique. 
I didn't know about point 3, though, so nevermind :)

dfialho replied:
Yes, you're right: main.py is not meant to be used as a module and should not be imported. The use of that technique is redundant in this case. Nevertheless, I believe it is still good practice to use it.

ric2b commented on mutual_debt/tests/test_mutual_debt_simplification.py:
```diff
@@ -1,5 +1,7 @@
 import unittest
-from mutual_debt_simplification import simplify_debt_graph, debt_list_to_graph
+
+from mutual_debt.mutual_debt_simplification import simplify_debt_graph, \
```
`mutual_debt.mutual_debt_simplification` is a bit redundant, the module could be changed to `simplification`

dfialho replied:
Totally agree

ric2b commented on setup.py:
```diff
@@ -0,0 +1,25 @@
+
+    packages=find_packages(exclude=['*.tests']),
+
+    install_requires=['graphviz==0.8'],
```
I didn't test it but I think this is just python bindings, you still need to install graphviz through the package manager for the script to work, no? If so, leaving a message after the installation finishes would be nice.

dfialho replied:
No. If you either install the package using pip or just by running `setup.py install`, it will install graphviz version 0.8 every time.

ric2b replied:
I just tested it and it does fail because the python package is just a wrapper/bindings/whatever for the actual package:
`graphviz.backend.ExecutableNotFound: failed to execute ['dot', '-Tpdf', '-O', 'Initial_Mutual_Debt.gv'], make sure the Graphviz executables are on your systems' PATH`

Fixed by installing the actual package: `sudo apt install graphviz`

dfialho replied:
I had no issue with this. I installed the application on a clean virtual environment and I was able to use it without any issue, including the generation of the "debt graph", which requires the graphviz package. How did you install the package?

Try installing with:
     
    pip3 install git+https://github.com/ric2b/Mutual_Debt_Simplification.git@master

dfialho replied:
My mistake. You are right: the system package *graphviz* must be installed for the program to be able to render the graph to a PDF

ric2b commented on setup.py:
```diff
@@ -0,0 +1,25 @@
+        'test': ['unittest'],
+    },
+
+    entry_points={
```
really cool :)

ric2b replied:
I had some trouble with getting it to work because of Path issues on Ubuntu, it's fixed by adding `export PATH=~/.local/bin:$PATH` to ~/.bashrc and running `source ~/.bashrc` to activate it in the current session

dfialho replied:
I had the same issue, but I was using z-shell. I though that was the reason for '~/.local/bin' not be included in PATH. I don't understand why is this not included by default, since all python console scripts are stored there, when installed in the user directory.

ric2b reviewed (approved):
Nice work :)
See if you agree with my comments before merging.

Commit:
Rename module 'mutual_debt_simplification' to 'simplification'

Commit:
Rename module 'data_structures' to 'graph'

Commit:
Define a default file for the input data file

The <data_file> input argument is not optional, allowing users to
execute the application with no arguments if it uses the default
filename.

davidfialho14 commented:
I have made changes to accommodate the issues and suggestions in your comments
